bin/liquidator/src/rebalance.rs

```diff
@@ -423,6 +409,301 @@ impl Rebalancer {
+        }
+
+        let mut account = account.clone();
+        let create_or_replace_ixs = self
+            .mango_client
+            .serum3_create_or_replace_account_instruction(&mut account, *market_index, side)
+            .await?;
+        let cancel_ixs =
+            self.mango_client
+                .serum3_cancel_all_orders_instruction(&account, *market_index, 10)?;
+        let place_order_ixs = self
+            .mango_client
+            .serum3_place_order_instruction(
+                &account,
+                *market_index,
+                side,
+                limit_price,
+                max_base_lots,
+                ((limit_price * max_base_lots * market.pc_lot_size) as f64 * 1.01) as u64,
+                Serum3SelfTradeBehavior::CancelProvide,
+                if self.config.limit_order_distance_from_oracle_price_bps == 0 {
+                    Serum3OrderType::PostOnly
+                } else {
+                    Serum3OrderType::Limit
+                },
+                SystemTime::now().duration_since(UNIX_EPOCH)?.as_millis() as u64,
+                10,
+            )
+            .await?;
+
+        let seq_check_ixs = self
+            .mango_client
+            .sequence_check_instruction(&self.mango_account_address, &account)
+            .await?;
+
+        let mut ixs = PreparedInstructions::new();
+        ixs.append(create_or_replace_ixs);
+        ixs.append(cancel_ixs);
+        ixs.append(place_order_ixs);
+        ixs.append(seq_check_ixs);
+
+        let txsig = self
+            .mango_client
+            .send_and_confirm_owner_tx(ixs.to_instructions())
+            .await?;
+
+        info!(
+            %txsig,
+            "placed order for {} {} at price = {}",
+            token.native_to_ui(I80F48::from(native_amount)),
+            token.name,
+            limit_price,
+        );
+        if !self.refresh_mango_account_after_tx(txsig).await? {
+            return Ok(());
+        }
+
+        Ok(())
+    }
+
+    async fn settle_and_close_all_openbook_orders(&self) -> anyhow::Result<()> {
+        let account = self.mango_account()?;
+
+        for x in Self::shuffle(account.active_serum3_orders()) {
+            let token = self.mango_client.context.token(x.base_token_index);
+            let quote = self.mango_client.context.token(x.quote_token_index);
+            let market_index = x.market_index;
+            let market = self
+                .mango_client
+                .context
+                .serum3_markets
+                .get(&market_index)
+                .expect("no openbook market found");
+            self.settle_and_close_openbook_orders(&account, token, &market_index, market, quote)
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn settle_and_close_openbook_orders(
+        &self,
+        account: &Box<MangoAccountValue>,
+        token: &TokenContext,
+        market_index: &Serum3MarketIndex,
+        market: &Serum3MarketContext,
```

Comment:
separate thought: we should add the market_index to the Serum3MarketContext, TokenContext also contains the token_index, it's convenient